File: algorithms/0057.insert_interval/insert_interval.rs
impl Solution {
    pub fn insert(intervals: Vec<Vec<i32>>, new_interval: Vec<i32>) -> Vec<Vec<i32>> {
        let mut intervals = intervals;
        let mut i = 0;
        while i < intervals.len() && intervals[i][0] < new_interval[0] {
            i += 1;
        }
        intervals.insert(i, new_interval);

        let mut i = 0;
        let mut insert_intervals = vec![];
        while i < intervals.len() {
            let (first, mut last) = (intervals[i][0], intervals[i][1]);
            while i + 1 < intervals.len() && last + 1 > intervals[i + 1][0] {
                last = last.max(intervals[i + 1][1]);
                i += 1;
            }
            insert_intervals.push(vec![first, last]);
            i += 1;
        }
        insert_intervals
    }
}

File: algorithms/0057.insert_interval/insert_interval.cpp
class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        bool inserted = false;
        for (auto it = intervals.begin(); it != intervals.end(); ++it) {
            if ((*it)[0] >= newInterval[0]) {
                intervals.insert(it, newInterval);
                inserted = true;
                break;
            }
        }
        if (!inserted) {
            intervals.emplace_back(newInterval);
        }
        vector<vector<int>> ans;
        ans.emplace_back(intervals[0]);
        for (size_t i = 1; i < intervals.size(); ++i) {
            auto &cur = intervals[i];
            if (ans.back()[1] >= cur[0]) {
                ans.back()[1] = max(ans.back()[1], cur[1]);
            } else {
                ans.emplace_back(cur);
            }
        }
        return ans;
    }
};